JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_BinaryStream.c

```diff
@@ -0,0 +1,286 @@
+						if (m_Data[index] != 0)
+						{
+							// The value is at the start of a block of 4 bytes
+							if (((index - lastZero) % 4) == 0)
+							{
+								m_NonZeroes.Insert(index);
+								lastZero = index + 1;
+							}
+						}
+					}
+					
+					index++;
+				}
+
+				// Don't need all non-zero variables
+				m_NonZeroes.Resize(m_Data.Count() % 4);
+				m_NonZeroesDirty = false;
+
+				CloseFile(fileHandle);
+				
+				break;
+			}
+			case FileMode.APPEND:
+			case FileMode.WRITE:
+			{
+				UpdateNonZeroFields();
+				
+				CloseFile(fileHandle);
+				
+				FileSerializer fs = new FileSerializer();
+				fs.Open(path, FileMode.APPEND);
+
+				int ivalue;
+
+				int zeroCheck = 0;
+				while (index < m_Data.Count())
+				{
+					if (zeroCheck < 4 && m_NonZeroes[zeroCheck] == index)
+					{
+						fs.Close();
+						fileHandle = OpenFile(path, FileMode.APPEND);
+						
+						ivalue = m_Data[index++];
+						FPrint(fileHandle, ivalue.AsciiToString());
+						zeroCheck++;
+						
+						CloseFile(fileHandle);
+						fs.Open(path, FileMode.APPEND);
+						continue;
+					}
+
+					int b0 = m_Data[index++];
+					int b1 = m_Data[index++];
+					int b2 = m_Data[index++];
+					int b3 = m_Data[index++];
+
+					ivalue = ((b3) << 24) + ((b2) << 16) + ((b1) << 8) + (b0);
+					fs.Write(ivalue);
+				}
+				
+				fs.Close();
+
+				break;
+			}
+		}
+		return true;
+	}
+	
+	void UpdateNonZeroFields()
+	{
+		if (!m_NonZeroesDirty) return;
+		
+		m_NonZeroes.Clear();
+		
+		int needed = m_Data.Count() % 4;
+		int lastZero = 0;
+		
+		for (int index = 0; index < m_Data.Count(); index++)
+		{
+			if (m_NonZeroes.Count() >= needed) break;
+
+			if (m_Data[index] == 0) continue;
+
+			// The value is at the start of a block of 4 bytes
+			if (((index - lastZero) % 4) == 0)
+			{
+				m_NonZeroes.Insert(index);
+				lastZero = index + 1;
+			}
+		}
+
+		m_NonZeroesDirty = false;
+	}
+
+	override void Write(int value)
+	{
+		m_Data.InsertAt(value, m_Position++);
+		
+		m_NonZeroesDirty = true;
+	}
+
+	override void WriteChar(string value)
+	{
+		Write(value.Hash());
+	}
+
+	override void WriteBool(bool value)
+	{
+		if (value) Write(0x00000001);
+		else Write(0x00000000);
+	}
+
+	override void WriteInt(int value)
+	{
+		Write((value >> 24) & 0x000000FF);
+		Write((value >> 16) & 0x000000FF);
+		Write((value >> 8 ) & 0x000000FF);
+		Write((value	  ) & 0x000000FF);
+	}
+
+	override void WriteFloat(float value)
+	{
+		WriteInt(CF_Cast<float, int>.Reinterpret(value));
+	}
+
+	override void WriteVector(vector value)
+	{
+		WriteFloat(value[0]);
+		WriteFloat(value[1]);
+		WriteFloat(value[2]);
+	}
+
+	override void WriteString(string value)
+	{
+		WriteInt(value.Length());
+		for (int i = 0; i < value.Length(); i++)
+		{
+			Write(value[i].Hash());
+		}
+	}
+
+	override void WriteCString(string value)
+	{
+		for (int i = 0; i < value.Length(); i++)
+		{
+			Write(value[i].Hash());
+		}
+		Write(0);
+	}
+
+	override int Read()
```

Comment:
Read ... what? A byte, if so this should be ReadByte()

Otherwise causes confusion with the other ReadX methods.